### PR TITLE
chore(liveness): allow selecting all cameras, allow camera selection …

### PR DESCRIPTION
**Pull request by thaddmt**
…on mobile

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 1]
```diff
@@ -115,6 +115,8 @@ export const LivenessCameraModule = (
 
   const isFaceMovementChallenge =
     useLivenessSelector(selectChallengeType) === FACE_MOVEMENT_CHALLENGE.type;
+  const isMobileFaceMovementAndLightChallenge =
```
Comment (reesscot) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:118]: Can we flip this logic to be more readable, something like 
`const allowSelectableDevices = !isMobile || (isMobile && isFaceMovementChallenge)`
then add a comment why we only support desktop or mobile and facemovement?

File: packages/e2e/features/ui/components/liveness/disable-start-screen.feature [hunk 2]
```diff
@@ -24,6 +24,3 @@ Feature: Disable Start Screen
       Then I click the "FaceMovementAndLightChallenge" selectfield and select the "FaceMovementChallenge" option
       Then I see "FaceMovementChallenge"
       Then I see "liveness-detector" element
-      Then I see the "Face didn't fit inside oval in time limit." timeout error
```
Comment (esauerbo) [inline packages/e2e/features/ui/components/liveness/disable-start-screen.feature:27]: Why do these need to be removed?

Comment (thaddmt) [reply packages/e2e/features/ui/components/liveness/disable-start-screen.feature:27]: Rekognition updated the values for face distance which make it so these tests no longer work, only on the no light stuff, figure that's unrelated to this PR but just want to unblock this

Comment (esauerbo) [reply packages/e2e/features/ui/components/liveness/disable-start-screen.feature:27]: Oh gotcha, what updates would we need to get them passing? But make sense as a follow up

Comment (thaddmt) [reply packages/e2e/features/ui/components/liveness/disable-start-screen.feature:27]: We would need to update the video we use in the e2e test haha

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 3]
```diff
@@ -115,6 +115,8 @@
 
   const isFaceMovementChallenge =
     useLivenessSelector(selectChallengeType) === FACE_MOVEMENT_CHALLENGE.type;
+  const allowSelectableDevicesMobile =
+    !isMobileScreen || (isMobileScreen && isFaceMovementChallenge);
```
[the remark this answers is not in the document]
Comment (reesscot) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx]: Why does this think isMobileScreen is always true?

[the remark this answers is not in the document]
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx]: I believe because if `isMobileScreen === false` then the left hand side will always short circuit the whole thing to be true. So you only evaluate the right side if `isMobileScreen === true` so it's redundant

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 4]
```diff
@@ -116,7 +116,7 @@ export const LivenessCameraModule = (
   const isFaceMovementChallenge =
     useLivenessSelector(selectChallengeType) === FACE_MOVEMENT_CHALLENGE.type;
   const allowSelectableDevicesMobile =
```
Comment (reesscot) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:118]: This is controlling whether we allow selectableDevices regardless of whether it's mobile or not, right? So should this just be `allowSelectableDevices`?

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx]: boolean algebra 🙃

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 5]
```diff
@@ -116,7 +116,7 @@ export const LivenessCameraModule = (
...
-    !isMobileScreen || (isMobileScreen && isFaceMovementChallenge);
+    !isMobileScreen || isFaceMovementChallenge;
```
Comment (reesscot) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:119]: If I'm reading this right, this will allow all cameras on all desktop for any challenge, and then for  faceMovementChallenge for desktop or mobile. Is that correct?

Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx]: There isn't a unit test specifically for camera selection, i can add it

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx (hunk 4 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx]: Update the names to be a little more clear and added a comment

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 6]
```diff
@@ -160,6 +160,15 @@ export const LivenessCameraModule = (
     videoWidth && videoHeight ? videoWidth / videoHeight : 0
   );
 
+  // Only allow on mobile screen if it is no light challenge
+  const allowSelectableDevicesMobile =
+    !isMobileScreen || isFaceMovementChallenge;
+  const allowSelectableDevices =
+    isStartView &&
+    allowSelectableDevicesMobile &&
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:168]: If i'm reading correctly we only allow device selection for mobile?

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 7]
```diff
@@ -418,38 +427,33 @@ export const LivenessCameraModule = (
...
-            !isMobileScreen &&
-            selectableDevices &&
-            selectableDevices.length > 1 && (
-              <Flex className={LivenessClassNames.StartScreenCameraSelect}>
-                <View
-                  className={
-                    LivenessClassNames.StartScreenCameraSelectContainer
-                  }
+          {allowSelectableDevices && (
+            <Flex className={LivenessClassNames.StartScreenCameraSelect}>
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:431]: Let's make a standalone component to get the file size down here a bit

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx (hunk 6 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:168]: We allow it on mobile only if it is a `FaceMovementChallenge` only (no light)

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+import React from 'react';
+import { LivenessClassNames } from '../types/classNames';
+
+interface CameraSelectorProps {
+  selectedDeviceId?: string;
+  onCameraChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
+  selectableDevices: MediaDeviceInfo[];
+}
+
+export const CameraSelector = (props: CameraSelectorProps): JSX.Element => {
```
Comment (reesscot) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx:11]: nice!

File: packages/react-liveness/src/components/FaceLivenessDetector/utils/helpers.ts [hunk 9]
```diff
@@ -8,5 +8,4 @@ export const STATIC_VIDEO_CONSTRAINTS: MediaTrackConstraints = {
     ideal: 480,
   },
   frameRate: { min: 15, ideal: 30, max: 30 },
-  facingMode: 'user',
```
Comment (reesscot) [inline packages/react-liveness/src/components/FaceLivenessDetector/utils/helpers.ts:11]: Just to confirm, this is no longer needed because on no light challenge we only show the dropdown on desktop, and on desktop the facing mode is not relevant?

**Review by reesscot — DISMISSED** [review]

File: packages/react-liveness/src/components/FaceLivenessDetector/utils/helpers.ts (hunk 9 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/utils/helpers.ts:11]: Exactly, desktop cameras do not have a facingMode value

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        </Label>
+        <SelectField
+          id="amplify-liveness-camera-select"
+          testId="amplify-liveness-camera-select"
+          label="Camera"
+          labelHidden
+          value={selectedDeviceId}
+          onChange={onCameraChange}
+        >
+          {selectableDevices?.map((device) => (
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx:30]: Since we are gating rendering of this component think we can remove the optional chaining here

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 11]
```diff
@@ -160,6 +153,15 @@ export const LivenessCameraModule = (
     videoWidth && videoHeight ? videoWidth / videoHeight : 0
   );
 
+  // Only allow on mobile screen if it is no light challenge
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:156]:
```suggestion
  // Only mobile device camera selection for no light challenge
```

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 12]
```diff
@@ -418,38 +420,13 @@ export const LivenessCameraModule = (
...
-                </View>
-              </Flex>
-            )}
+          {allowSelectableDevices && (
+            <CameraSelector
+              onCameraChange={onCameraChange}
+              selectableDevices={selectableDevices}
+              selectedDeviceId={selectedDeviceId}
+            />
+          )}
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:429]: Prefer ternary returning `null` as the second condition in place of returning `false`

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx [hunk 13]
```diff
@@ -160,6 +153,15 @@ export const LivenessCameraModule = (
     videoWidth && videoHeight ? videoWidth / videoHeight : 0
   );
 
+  // Only allow on mobile screen if it is no light challenge
+  const allowSelectableDevicesMobile =
+    !isMobileScreen || isFaceMovementChallenge;
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:158]: It's a bit hard to follow that `allowSelectableDevicesMobile` is `true` when `isMobileScreen` is `false`. Realize that it going to be hard for this not to be somewhat confusing but maybe we can update the approach to something like:
```ts
const hasMultipleDevices = !!selectableDevices?.length;
const allowDeviceSelection = (isStartView && hasMultipleDevices) && (!isMobileScreen || isFaceMovementChallenge)
```

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx [hunk 14]
```diff
@@ -0,0 +1,39 @@
+import { Flex, Label, SelectField, View } from '@aws-amplify/ui-react';
+import React from 'react';
+import { LivenessClassNames } from '../types/classNames';
+
+interface CameraSelectorProps {
+  selectedDeviceId?: string;
+  onCameraChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
+  selectableDevices: MediaDeviceInfo[];
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx:8]: IMO we can simplify the naming of the `props` a bit here since camera selection is implied by the component name, e.g.
```ts
  deviceId?: string;
  onSelect?: (e: React.ChangeEvent<HTMLSelectElement>) => void;
  devices: MediaDeviceInfo[];
```

**Review by calebpollman — COMMENTED** [review]
Minor feedback, lmk your thoughts!

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx (hunk 10 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx:30]: Good catch, removed

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx (hunk 12 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:429]: fixed

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx (hunk 13 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/LivenessCameraModule.tsx:158]: Think that makes sense, think this is much cleaner

File: packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx (hunk 14 above)
Comment (thaddmt) [reply packages/react-liveness/src/components/FaceLivenessDetector/LivenessCheck/CameraSelector.tsx:8]: on board with this

**Review by calebpollman — APPROVED** [review]

**Review by reesscot — APPROVED** [review]

**Review by calebpollman — APPROVED** [review]
